lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,241 @@
+require 'set'
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # Files present in :files are not processed further. The others
+      #   are added to the Hash. This behavior is the same implemented in openstack-infra/shade
+      #   see https://github.com/openstack-infra/shade/blob/1d16f64fbf376a956cafed1b3edd8e51ccc16f2c/shade/openstackcloud.py#L1200
+
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :template_ori
+        attr_reader :visited
+        attr_reader :max_files_size
+
+        def initialize(template, files = nil)
+          @template = template
+          @template_ori = template
+          @template_base_url = nil
+          @files = files || {}
+          @visited = Set.new
+        end
+
+        def get_files
+          Fog::Logger.debug("Processing template #{@template}")
+          @template = get_template_contents(@template)
+          Fog::Logger.debug("Template processed. Populated #{@files}")
+          @files
+        end
+
+        # Return string
+        def url_join(prefix, suffix)
+          if prefix
+            prefix += '/' if prefix && !prefix.to_s.end_with?("/")
+            suffix = URI.join(prefix, suffix)
+            # Force URI to use traditional file scheme representation.
+            suffix.host = "" if suffix.scheme == "file"
+            suffix = suffix.to_s
```

Comment:
don't need this line

Reply:
@bzwei I did it to ensure returned value is a string and not an uri.